pkg/features/kube_features.go

```diff
@@ -23,1037 +23,1040 @@ import (
 
-	genericfeatures.APIListChunking: {Default: true, PreRelease: featuregate.Beta},
+// owner: @cofyc
+// alpha: v1.21
+const VolumeCapacityPriority featuregate.Feature = "VolumeCapacityPriority"
 
-	genericfeatures.APIPriorityAndFairness: {Default: true, PreRelease: featuregate.Beta},
+var _ = add(VolumeCapacityPriority, featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha})
 
-	genericfeatures.APIResponseCompression: {Default: true, PreRelease: featuregate.Beta},
+// owner: @ksubrmnn
+// alpha: v1.14
+//
+// Allows kube-proxy to create DSR loadbalancers for Windows
+const WinDSR featuregate.Feature = "WinDSR"
 
-	genericfeatures.AdvancedAuditing: {Default: true, PreRelease: featuregate.GA},
+var _ = add(WinDSR, featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha})
 
-	genericfeatures.CustomResourceValidationExpressions: {Default: false, PreRelease: featuregate.Alpha},
+// owner: @ksubrmnn
+// alpha: v1.14
+// beta: v1.20
+//
+// Allows kube-proxy to run in Overlay mode for Windows
+const WinOverlay featuregate.Feature = "WinOverlay"
 
-	genericfeatures.DryRun: {Default: true, PreRelease: featuregate.GA},
+var _ = add(WinOverlay, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
 
-	genericfeatures.OpenAPIEnums: {Default: true, PreRelease: featuregate.Beta},
+// owner: @marosset
+// alpha: v1.22
+// beta: v1.23
+//
+// Enables support for 'HostProcess' containers on Windows nodes.
+const WindowsHostProcessContainers featuregate.Feature = "WindowsHostProcessContainers"
 
-	genericfeatures.OpenAPIV3: {Default: true, PreRelease: featuregate.Beta},
+var _ = add(WindowsHostProcessContainers, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
 
-	genericfeatures.ServerSideApply: {Default: true, PreRelease: featuregate.GA},
+func init() {
+	runtime.Must(utilfeature.DefaultMutableFeatureGate.Add(defaultKubernetesFeatureGates))
+}
 
-	genericfeatures.ServerSideFieldValidation: {Default: false, PreRelease: featuregate.Alpha},
+// defaultKubernetesFeatureGates consists of all known Kubernetes-specific feature keys.
+// To add a new feature, define a key for it above and add it here. The features will be
+// available throughout Kubernetes binaries.
+//
+// Entries are separated from each other with blank lines to avoid sweeping gofmt changes
+// when adding or removing one entry.
+var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureSpec{}
 
-	// features that enable backwards compatibility but are scheduled to be removed
-	// ...
-	HPAScaleToZero: {Default: false, PreRelease: featuregate.Alpha},
+func add(feature featuregate.Feature, spec featuregate.FeatureSpec) bool {
+	defaultKubernetesFeatureGates[feature] = spec
+	return true
 }
+
+// inherited features from generic apiserver, relisted here to get a conflict if it is changed
+// unintentionally on either side:
```

Comment:
nit: It would be tidier to juts put these in an init block.

Reply:
Done.

I also rebased because of a conflict with https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#diff-71e3b98f9a6bbf5b8421e26a7ba0c079f397cd8d49abacdad943c66a4f44f03d

The diff of just this file since my initial push is:

```patch
$ git diff pohly/feature-gates-sorting pkg/features/kube_features.go
diff --git a/pkg/features/kube_features.go b/pkg/features/kube_features.go
index f8e94736944..a3ca86c423a 100644
--- a/pkg/features/kube_features.go
+++ b/pkg/features/kube_features.go
@@ -721,6 +721,16 @@ const NetworkPolicyStatus featuregate.Feature = "NetworkPolicyStatus"
 
 var _ = add(NetworkPolicyStatus, featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha})
 
+// owner: @kerthcet
+// kep: http://kep.k8s.io/3094
+// alpha: v1.25
+//
+// Allow users to specify whether to take nodeAffinity/nodeTaint into consideration when
+// calculating pod topology spread skew.
+const NodeInclusionPolicyInPodTopologySpread featuregate.Feature = "NodeInclusionPolicyInPodTopologySpread"
+
+var _ = add(NodeInclusionPolicyInPodTopologySpread, featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha})
+
 // owner: @xing-yang @sonasingh46
 // kep: http://kep.k8s.io/2268
 // alpha: v1.24
@@ -1017,7 +1027,27 @@ const WindowsHostProcessContainers featuregate.Feature = "WindowsHostProcessCont
 
 var _ = add(WindowsHostProcessContainers, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
 
+// Do not add new features at the end unless that is where it belongs due to
+// alphabetical sorting. See comment at the top for details.
+
 func init() {
+       // inherited features from generic apiserver, relisted here to get a conflict if it is changed
+       // unintentionally on either side:
+       add(genericfeatures.APIListChunking, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
+       add(genericfeatures.APIPriorityAndFairness, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
+       add(genericfeatures.APIResponseCompression, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
+       add(genericfeatures.AdvancedAuditing, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.GA})
+       add(genericfeatures.CustomResourceValidationExpressions, featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha})
+       add(genericfeatures.DryRun, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.GA})
+       add(genericfeatures.OpenAPIEnums, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
+       add(genericfeatures.OpenAPIV3, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
+       add(genericfeatures.ServerSideApply, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.GA})
+       add(genericfeatures.ServerSideFieldValidation, featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha})
+
+       // features that enable backwards compatibility but are scheduled to be removed
+       // ...
+       add(HPAScaleToZero, featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha})
+
        runtime.Must(utilfeature.DefaultMutableFeatureGate.Add(defaultKubernetesFeatureGates))
 }
 
@@ -1033,30 +1063,3 @@ func add(feature featuregate.Feature, spec featuregate.FeatureSpec) bool {
        defaultKubernetesFeatureGates[feature] = spec
        return true
 }
-
-// inherited features from generic apiserver, relisted here to get a conflict if it is changed
-// unintentionally on either side:
-
-var _ = add(genericfeatures.APIListChunking, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
-
-var _ = add(genericfeatures.APIPriorityAndFairness, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
-
-var _ = add(genericfeatures.APIResponseCompression, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
-
-var _ = add(genericfeatures.AdvancedAuditing, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.GA})
-
-var _ = add(genericfeatures.CustomResourceValidationExpressions, featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha})
-
-var _ = add(genericfeatures.DryRun, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.GA})
-
-var _ = add(genericfeatures.OpenAPIEnums, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
-
-var _ = add(genericfeatures.OpenAPIV3, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
-
-var _ = add(genericfeatures.ServerSideApply, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.GA})
-
-var _ = add(genericfeatures.ServerSideFieldValidation, featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha})
-
-// features that enable backwards compatibility but are scheduled to be removed
-// ...
-var _ = add(HPAScaleToZero, featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha})
```

Reply:
I added the "Do not add new features at the end unless that is where it belongs due to alphabetical sorting." comment at the end because that other PR ignored sorting and added at the end. Old habits die hard...

Looks like we need a verify script after all? I can add one, but let's do that in a separate PR and instead get this merged quickly. The risk of painful code conflicts is high.

Reply:
Yep, one more rebase was already needed.

Reply:
The corresponding change in my branch for https://github.com/kubernetes/kubernetes/pull/108218 in my branch is:

```patch
diff --git a/pkg/features/kube_features.go b/pkg/features/kube_features.go
index a3ca86c423a..08a537f343b 100644
--- a/pkg/features/kube_features.go
+++ b/pkg/features/kube_features.go
@@ -214,16 +214,6 @@ const CSIStorageCapacity featuregate.Feature = "CSIStorageCapacity"
 
 var _ = add(CSIStorageCapacity, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.GA, LockToDefault: true}) // remove in 1.26
 
-// owner: @huffmanca, @dobsonj
-// alpha: v1.19
-// beta: v1.20
-// GA: v1.23
-//
-// Determines if a CSI Driver supports applying fsGroup.
-const CSIVolumeFSGroupPolicy featuregate.Feature = "CSIVolumeFSGroupPolicy"
-
-var _ = add(CSIVolumeFSGroupPolicy, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.GA, LockToDefault: true}) // remove in 1.25
-
 // owner: @fengzixu
 // alpha: v1.21
 //
@@ -241,16 +231,6 @@ const CSRDuration featuregate.Feature = "CSRDuration"
 
 var _ = add(CSRDuration, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.GA, LockToDefault: true}) // remove in 1.26
 
-// owner: @gnufied
-// alpha: v1.18
-// beta: v1.20
-// GA: v1.23
-// Allows user to configure volume permission change policy for fsGroups when mounting
-// a volume in a Pod.
-const ConfigurableFSGroupPolicy featuregate.Feature = "ConfigurableFSGroupPolicy"
-
-var _ = add(ConfigurableFSGroupPolicy, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.GA, LockToDefault: true}) // remove in 1.25
-
 // owner: @jiahuif
 // alpha: v1.21
 // beta:  v1.22
@@ -787,15 +767,6 @@ const PodDeletionCost featuregate.Feature = "PodDeletionCost"
 
 var _ = add(PodDeletionCost, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Beta})
 
-// owner: @mortent
-// alpha: v1.3
-// beta:  v1.5
-//
-// Enable all logic related to the PodDisruptionBudget API object in policy
-const PodDisruptionBudget featuregate.Feature = "PodDisruptionBudget"
-
-var _ = add(PodDisruptionBudget, featuregate.FeatureSpec{Default: true, PreRelease: featuregate.GA, LockToDefault: true}) // remove in 1.25
-
 // owner: @egernst
 // alpha: v1.16
 // beta: v1.18
```

This shows the main advantage of reorganizing the file like this: each feature gate removal is done in one location.